<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>netmap.downstream package &#8212; Netmap Jan, 2026 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/jquery.js?v=5d32c60e"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="_static/documentation_options.js?v=f7525948"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="netmap-downstream-package">
<h1>netmap.downstream package<a class="headerlink" href="#netmap-downstream-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-netmap.downstream.clustering">
<span id="netmap-downstream-clustering-module"></span><h2>netmap.downstream.clustering module<a class="headerlink" href="#module-netmap.downstream.clustering" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.clustering.downstream_recipe">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.clustering.</span></span><span class="sig-name descname"><span class="pre">downstream_recipe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="headerlink" href="#netmap.downstream.clustering.downstream_recipe" title="Link to this definition"></a></dt>
<dd><p>Downstream reciepe for an LRP anndata object:
TODO: replace the config dict, to pass values</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.clustering.process">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.clustering.</span></span><span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spectral'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.clustering.process" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.clustering.spectral_clustering">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.clustering.</span></span><span class="sig-name descname"><span class="pre">spectral_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spectral'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.clustering.spectral_clustering" title="Link to this definition"></a></dt>
<dd><p>Run sklearn spectral clustering on the neighbour matrix in the anndata object.</p>
<p>Args:
adata: Anndata object
n_clu: Number of clusters to compute
key_added: The key to add the new labelling to [Default: spectral]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.clustering.unify_group_labelling">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.clustering.</span></span><span class="sig-name descname"><span class="pre">unify_group_labelling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.clustering.unify_group_labelling" title="Link to this definition"></a></dt>
<dd><p>Adjust group labelling such that grn_adata has the same group label than the
corresponding column in adata based on the grn column.</p>
<p>Returns the data objects and a score of the matching as the cost of the matching
divided by the number of cells.</p>
</dd></dl>

</section>
<section id="netmap-downstream-downstreammain-module">
<h2>netmap.downstream.downstreammain module<a class="headerlink" href="#netmap-downstream-downstreammain-module" title="Link to this heading"></a></h2>
</section>
<section id="module-netmap.downstream.edge_selection">
<span id="netmap-downstream-edge-selection-module"></span><h2>netmap.downstream.edge_selection module<a class="headerlink" href="#module-netmap.downstream.edge_selection" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.edge_selection.add_top_edge_annotation_cluster">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.edge_selection.</span></span><span class="sig-name descname"><span class="pre">add_top_edge_annotation_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.1]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_fill</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spectral'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.edge_selection.add_top_edge_annotation_cluster" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.edge_selection.add_top_edge_annotation_global">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.edge_selection.</span></span><span class="sig-name descname"><span class="pre">add_top_edge_annotation_global</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.1]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_fill</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'global'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.edge_selection.add_top_edge_annotation_global" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.edge_selection.data_preprr">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.edge_selection.</span></span><span class="sig-name descname"><span class="pre">data_preprr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">global_counter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_metadata_np</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_edges_val</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.edge_selection.data_preprr" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-netmap.downstream.final_downstream">
<span id="netmap-downstream-final-downstream-module"></span><h2>netmap.downstream.final_downstream module<a class="headerlink" href="#module-netmap.downstream.final_downstream" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.compute_edge_overlaps_simple">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">compute_edge_overlaps_simple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#netmap.downstream.final_downstream.compute_edge_overlaps_simple" title="Link to this definition"></a></dt>
<dd><p>Compute the percentage of overlapping edges between a GRN and multiple reference networks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – AnnData object with edges in <cite>grn_adata.var</cite>, containing ‘source’ and ‘target’ columns.</p></li>
<li><p><strong>net_list</strong> (<em>List</em><em>[</em><em>Tuple</em><em>[</em><em>str</em><em>, </em><em>pd.DataFrame</em><em>]</em><em>]</em>) – List of tuples (network_name, network_dataframe), where each dataframe
must include ‘source’ and ‘target’ columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary mapping each network name (with index suffix) to its overlap percentage.
Example: {‘STRING__0’: 0.32, ‘BioGRID__1’: 0.45}</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, float]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.create_regulon_activity_adata">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">create_regulon_activity_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_ucell</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_sources_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="headerlink" href="#netmap.downstream.final_downstream.create_regulon_activity_adata" title="Link to this definition"></a></dt>
<dd><p>Creates an AnnData object with regulon activity scores based on top GRN sources.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – Original GRN AnnData object, used for alignment of obs and embeddings.</p></li>
<li><p><strong>data_ucell</strong> (<em>pd.DataFrame</em>) – DataFrame containing UCell scores for all signatures.</p></li>
<li><p><strong>top_sources_list</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of top source names to include as regulons.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New AnnData object with regulon activity, aligned to grn_adata observations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.curate_network">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">curate_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tfs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.final_downstream.curate_network" title="Link to this definition"></a></dt>
<dd><p>Load CollecTRI data for a specific organism, format it, and curate the network data
based on the provided gene regulatory network (GRN) data.</p>
<section id="parameters">
<h3>Parameters:<a class="headerlink" href="#parameters" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a href="#id1"><span class="problematic" id="id2">grn_adata_</span></a><span class="classifier">AnnData</span></dt><dd><p>Annotated data object containing the gene regulatory network (GRN) data.</p>
</dd>
<dt>organism<span class="classifier">str, optional (default=”human”)</span></dt><dd><p>The organism to retrieve data for from CollecTRI.</p>
</dd>
<dt>tfs<span class="classifier">list of str or None, optional (default=None)</span></dt><dd><p>A list of transcription factors (source genes) to retain. If None, all TFs are used.</p>
</dd>
</dl>
</section>
<section id="returns">
<h3>Returns:<a class="headerlink" href="#returns" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>net_df_curated<span class="classifier">pandas.DataFrame</span></dt><dd><p>The curated network DataFrame, containing the merged GRN data and CollecTRI interaction data.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.filter_clusters_by_cell_count">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">filter_clusters_by_cell_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#netmap.downstream.final_downstream.filter_clusters_by_cell_count" title="Link to this definition"></a></dt>
<dd><p>Filter features (genes/edges) based on cell count differences between two clusters,
selecting the top fraction of features for each cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – AnnData object containing feature metadata in <cite>.var</cite>.
Must include columns <cite>cell_count_&lt;metric_tag&gt;_0</cite> and <cite>cell_count_&lt;metric_tag&gt;_1</cite>.</p></li>
<li><p><strong>metric_tag</strong> (<em>str</em>) – Identifier (suffix) for the cell count metric, e.g. ‘0.95’ → columns
<cite>cell_count_0.95_0</cite> and <cite>cell_count_0.95_1</cite>.</p></li>
<li><p><strong>top_fraction</strong> (<em>float</em>) – Fraction (0–1) of top features to retain per cluster.
For example, 0.2 retains the top 20% of features.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>filtered_adata</strong> – Subset of the input AnnData object containing only the top-selected features.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.filter_grn_by_top_signatures">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">filter_grn_by_top_signatures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_ucell</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_top_ranked</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'z_score'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#netmap.downstream.final_downstream.filter_grn_by_top_signatures" title="Link to this definition"></a></dt>
<dd><p>Filters a GRN (Gene Regulatory Network) AnnData object to keep only the edges
corresponding to the top-ranked signatures per cluster based on UCell scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_ucell</strong> (<em>pd.DataFrame</em>) – DataFrame containing UCell scores for each signature across cells, with a
‘spectral’ column indicating cluster assignments.</p></li>
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – The GRN AnnData object to filter, containing edge metadata in <cite>.var</cite>.</p></li>
<li><p><strong>keep_top_ranked</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of top signatures to retain per cluster (default is 100).</p></li>
<li><p><strong>filter_by</strong> (<em>str</em><em>, </em><em>optional</em>) – Metric used to rank signatures. Options include “P_Value”, “Adjusted_P_Value”, or “z_score” (default is “P_Value”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>Filtered AnnData object containing only edges corresponding to top-ranked signatures.</p></li>
<li><p>List of unique top source names used to filter the GRN.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[Optional[AnnData], List[str]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>For each cluster, signatures are ranked using the chosen metric.</p></li>
<li><p>When <cite>filter_by=’z_score’</cite>, a combined score of -log10(P_Value) and z_score is used.</p></li>
<li><p>The function returns <cite>None</cite> and an empty list if the input GRN has no variables.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.filter_signatures_by_Ucell">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">filter_signatures_by_Ucell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#netmap.downstream.final_downstream.filter_signatures_by_Ucell" title="Link to this definition"></a></dt>
<dd><p>Filters gene signatures by cluster and computes UCell scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – AnnData object containing GRN (gene regulatory network) information.</p></li>
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object containing gene expression counts in the ‘counts’ layer.</p></li>
<li><p><strong>ncores</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cores to use for parallel computation, by default 100.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with UCell scores merged with the ‘spectral’ cluster labels.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.get_top_targets">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">get_top_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_inter_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_per_source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">750</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spectral'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reg_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.final_downstream.get_top_targets" title="Link to this definition"></a></dt>
<dd><p>Selects top gene targets per source from a clustered gene interaction AnnData.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_inter_adata</strong> (<em>AnnData</em>) – Gene interaction AnnData with <cite>var</cite> containing ‘source’ and ‘target’.</p></li>
<li><p><strong>adata</strong> (<em>AnnData</em>) – Expression AnnData for ranking genes.</p></li>
<li><p><strong>top_per_source</strong> (<em>int</em><em>, </em><em>default=750</em>) – Number of top targets to select per source.</p></li>
<li><p><strong>col_cluster</strong> (<em>str</em><em>, </em><em>default='spectral'</em>) – Column in obs defining clusters.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>gene_inter_adata_filtered</strong> (<em>AnnData</em>) – Filtered AnnData containing top edges.</p></li>
<li><p><strong>reglon_sizes</strong> (<em>list of int</em>) – Sizes of regulatory regions per source.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.plot_reg">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">plot_reg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regulon</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'network'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'hierarchical'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'network_plots/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.final_downstream.plot_reg" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.plot_reg_curated">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">plot_reg_curated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">net_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regulon</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'network'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'hierarchical'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'network_plots/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.final_downstream.plot_reg_curated" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.final_downstream.plot_shared_targets_heatmap">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.final_downstream.</span></span><span class="sig-name descname"><span class="pre">plot_shared_targets_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RdBu_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Clustered</span> <span class="pre">Heatmap</span> <span class="pre">of</span> <span class="pre">Shared</span> <span class="pre">Targets'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.final_downstream.plot_shared_targets_heatmap" title="Link to this definition"></a></dt>
<dd><p>Create a clustered heatmap showing shared target counts between sources.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grn_adata</strong> (<em>AnnData</em>) – An AnnData object where the gene regulatory network info is in <cite>grn_adata.var</cite>.
Must contain columns ‘source’ and ‘target’.</p></li>
<li><p><strong>genes</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of source genes to include. If None, all sources are used.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the heatmap figure.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) – Colormap for the heatmap.</p></li>
<li><p><strong>metric</strong> (<em>str</em><em>, </em><em>optional</em>) – Distance metric for clustering (passed to seaborn.clustermap).</p></li>
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em>) – Linkage method for clustering (passed to seaborn.clustermap).</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title for the heatmap.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>shared_target_matrix</strong> – A symmetric DataFrame of shared target counts between sources.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-netmap.downstream.gene_clustering">
<span id="netmap-downstream-gene-clustering-module"></span><h2>netmap.downstream.gene_clustering module<a class="headerlink" href="#module-netmap.downstream.gene_clustering" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.gene_clustering.add_clusters_to_adata">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.gene_clustering.</span></span><span class="sig-name descname"><span class="pre">add_clusters_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist_linkage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'correlation_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.gene_clustering.add_clusters_to_adata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.gene_clustering.compute_regression">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.gene_clustering.</span></span><span class="sig-name descname"><span class="pre">compute_regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cophenet_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist_linkage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.gene_clustering.compute_regression" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.gene_clustering.get_hierarchical_clustering">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.gene_clustering.</span></span><span class="sig-name descname"><span class="pre">get_hierarchical_clustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.gene_clustering.get_hierarchical_clustering" title="Link to this definition"></a></dt>
<dd><p>Performs hierarchical clustering</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>ad.AnnData</em>) – The input AnnData object.</p></li>
<li><p><strong>tree_threshold</strong> (<em>float</em>) – The maximum cophenet distance to include in the regression fit.</p></li>
<li><p><strong>correlation_threshold</strong> (<em>float</em>) – The minimum correlation value to define clusters.</p></li>
<li><p><strong>quantile</strong> (<em>float</em>) – The quantile for the regression model (e.g., 0.1 for 10th percentile).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.gene_clustering.plot_regression_and_dendrogram">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.gene_clustering.</span></span><span class="sig-name descname"><span class="pre">plot_regression_and_dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist_linkage</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.gene_clustering.plot_regression_and_dendrogram" title="Link to this definition"></a></dt>
<dd><p>Plots the data points, quantile regression line, automatic cutoff, and the dendrogram.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="netmap.downstream.gene_clustering.plot_scatter_plot">
<span class="sig-prename descclassname"><span class="pre">netmap.downstream.gene_clustering.</span></span><span class="sig-name descname"><span class="pre">plot_scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#netmap.downstream.gene_clustering.plot_scatter_plot" title="Link to this definition"></a></dt>
<dd><p>Plots the data points, to find the cutoff for which a linear model can be fit.</p>
</dd></dl>

</section>
<section id="module-netmap.downstream">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-netmap.downstream" title="Link to this heading"></a></h2>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Netmap</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules.html">netmap</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Anne Hartebrodt, Mhaned Oubounyt.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/netmap.downstream.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>